### PR TITLE
Chat loop prevention: hard rejection + anti-repeat retry ladder

### PR DESCRIPTION
**Pull request by holdenk**
## Summary

Implements a multi-layer defense against the observed production failure where chat models re-send their previous reply verbatim instead of using the user's new answer. The fix combines hard rejection of repeated candidates in scoring, an anti-repeat retry with explicit instruction and hotter sampling, and optional side-by-side alternate answers.

### Core Loop Prevention
- **Response similarity detection** (`ml/response_similarity.py`): New stdlib-only module with text-similarity helpers (exact/near-repeat detection via sequence matching and Jaccard similarity) shared by both the per-backend retry loop and candidate scoring without import cycles.
- **Hard rejection in scoring** (`chat/llm_client.py`): 
  - `find_repeated_reply()` detects when a candidate repeats recent assistant replies (covers A-A loops and A-B-A alternation)
  - `user_requested_repeat()` exempts explicit repeat requests from rejection
  - `is_canned_reply()` exempts mandated verbatim responses (e.g., Medicaid FAQ blocks)
  - Repeated candidates scored `-inf` so they never win in `best_within_timelimit()`
- **Anti-repeat retry ladder** (`chat/retry_handler.py`):
  - `ANTI_REPEAT_NOTE` instruction added to retry message when a repeat is detected
  - Temperature increased to 0.9 for hotter sampling on retry
  - Retry candidates still repeating get `REPEAT_LAST_RESORT_PENALTY` (finite but decisive) so a non-repeat always wins, but a repeat can still be delivered as absolute last resort
- **Per-backend self-heal** (`ml/ml_models.py`): `RemoteOpenLike.generate_chat_response()` detects near-repeats in fresh generations and retries once with corrective feedback before the caller's scoring ladder sees it

### Chat Interface Enhancements
- **State hint from IP** (`fhi_users/audit.py`): `guess_us_state()` derives best-effort US state from connection IP using geoip2fast (transient, unconfirmed, never persisted). Flows into LLM context each turn to help Medicaid questions start with "looks like you might be in California — is that right?"
- **Side-by-side alternates** (`chat_interface.py`): Runner-up answer from top-level LLM fan-out offered to client as collapsible alternate when presentable
- **Debug LLM input frame** (`chat_interface.py`): When `debug_llm=True` (gated to DEBUG deployments and staff), each turn sends a `debug_llm_input` frame showing exactly what was sent to the backend model
- **Terse-reply bridge note** (`chat_interface.py`): Short user messages (≤60 chars) right after an assistant question get a note explaining the model needs more detail — addresses the shape where models historically looped by re-asking instead of using terse answers

### Observability
- **Metrics** (`ml/ml_metrics.py`):
  - `fhi_chat_repeated_responses_total` with actions `rejected_candidates` (hard rejection in primary pass) and `delivered_repeat` (last-resort delivery)
  - `fhi_chat_alternate_answers_total` and `fhi_chat_answer_feedback_total` for side-by-side answer tracking
- **WebSocket feedback frame** (`websockets.py`): Lightweight `{"answer_feedback": {"preferred": ...}}` frame records which answer users preferred without starting an LLM turn

### Utility & Infrastructure
- **`best_two_within_timelimit()`** (`utils.py`): New variant of `best_within_timelimit()` that returns both the best and runner-up results (used to populate the alternate answer)
- **`-inf` scoring semantics** (`utils.py`): Hard-rejected results (scored `-inf`) are now treated as invalid and never returned, even if they complete first — guards against a bad-but-fast result winning by default
- **Context bounding** (`chat/context_manager.py`): `bound_summary_context()` trims accreted summary context with a `TRIMMED

https://claude.ai/code/session_01SMRUqEUViHv75a7Xns2AU1

## Summary by CodeRabbit

* **New Features**
  * Added automatic U.S. state and ASN detection when GeoIP data is available.
  * Added expandable alternate chat answers, preference feedback, and optional debugging views.
  * External model participation is now enabled by default and can be disabled.
* **Improvements**
  * Chat detects and recovers from repetitive responses.
  * Conversation summaries and stored history are bounded and better organized.
  * Internal context notes are hidden from user-facing history.
  * User messages are preserved more reliably during failures and replay.
* **Documentation**
  * Added setup, verification, security, and behavior guidance for GeoIP functionality.

### CODE REVIEW

File: fhi_users/audit.py [hunk 1]
```diff
@@ -336,31 +338,264 @@ def get_asn_info(ip_address: Optional[str]) -> tuple[str, str]:
...
-        # geoip2fast not installed, that's fine
-        pass
+        result = reader.lookup(str(ip_address).strip())
+        if result is None:
+            return ("", "")
+        # geoip2fast exposes asn_name (and asn_cidr); some builds/forks also
+        # expose a numeric asn. Probe defensively.
+        asn_num = str(getattr(result, "asn", "") or "")
+        asn_name = str(getattr(result, "asn_name", "") or "")[:200]
+        return (asn_num, asn_name)
```
[the remark this answers is not in the document]
Comment (holdenk) [reply fhi_users/audit.py:363]: Verified against geoip2fast 1.2.2 locally: the lookup result exposes `asn_name`/`asn_cidr` but no numeric ASN field at all, so there is no documented numeric source to restore — with this library the numeric slot was empty before this PR too (the old per-call reader probed the same nonexistent `result.asn`). Kept the contract as-is: `asn_name` is the meaningful field, the numeric probe stays for forks/builds that add one, and the comment in 8c7d36e now states this explicitly so it doesn't read as data loss.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: fighthealthinsurance/ml/ml_models.py [hunk 2]
```diff
@@ -1305,11 +1338,25 @@ async def generate_chat_response(
...
+                # priority over the missing-panda feedback -- a repeated
+                # reply is the worse failure.
+                result_extra = (
+                    "Your previous draft repeated your last reply, which the "
+                    "user has already seen. Do not repeat it -- respond to "
+                    "the user's newest message directly, using the details "
+                    "they have already provided.\n\n"
+                )
+                temperature = min(1.0, temperature + 0.15)
+            elif result and len(result) > 0 and "🐼" not in result:
```
[the remark this answers is not in the document]
Comment (holdenk) [reply fighthealthinsurance/ml/ml_models.py:1387]: Adopted the canned-reply half in 8c7d36e: the exemption helpers moved to the cycle-free `response_similarity` module and `_repeats_last_reply` now skips mandated canned replies (with a regression test).

Deliberately did NOT apply `user_requested_repeat` at this layer: `generate_chat_response` only sees the wrapped prompt, and the system-injected wrapper text (the anti-repeat note, the terse-reply bridge) itself contains the word "repeat", so the regex would false-positive and disable repeat detection exactly when it matters most. The scoring layer applies that exemption against the raw user message, so an explicit "please repeat that" costs at most one extra bounded serial call here and then flows through.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: fighthealthinsurance/chat/context_manager.py [hunk 3]
```diff
@@ -63,9 +83,14 @@ async def prepare_history_for_llm(
...
-    # Check if we should summarize (every N messages after threshold)
+    # Check if we should summarize (once per N-message band after the
+    # threshold). History normally grows by 2 per turn, but consecutive
+    # same-role messages get merged (ensure_message_alternation) so the
+    # length can change parity -- a strict `% N == 0` then never fires again
+    # and dropped messages stop being folded into the summary at all. `<= 1`
+    # fires exactly once per band for both parities.
     messages_over_threshold = len(history_for_llm) - DEFAULT_MESSAGES_TO_KEEP
-    should_summarize = messages_over_threshold % SUMMARIZATION_INTERVAL == 0
+    should_summarize = messages_over_threshold % SUMMARIZATION_INTERVAL <= 1
```
[the remark this answers is not in the document]
Comment (holdenk) [reply fighthealthinsurance/chat/context_manager.py:161]: Fixed the harmful half in 8c7d36e: `_summarize_history` now REPLACES a previous full-prefix summary block instead of nesting it (every call re-covers the whole dropped prefix, so the fresh summary supersedes the old one; non-summary context like microsite/PubMed additions is preserved). With that, the remainder-0-then-1 double trigger produces byte-identical output and its only cost is one redundant bounded call in the rare failed-turn-at-a-boundary case — tracking the exact summarized boundary would need persisted per-chat state, which that residual cost doesn't justify. The trigger comment now documents the tradeoff, and there's a regression test for the replacement behavior.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: fighthealthinsurance/chat_interface.py [hunk 4]
```diff
@@ -920,6 +1063,59 @@ def wrap_for_llm(text: str) -> str:
...
+        # treat chat.chat_history as the PRE-turn history (the current
+        # message travels separately). Guarded: persistence problems must
+        # not block the turn -- worst case we fall back to the old
+        # persist-at-end behavior.
+        if user_message and user_message.strip():
+            try:
+                await apersist_chat_turn(
+                    chat,
+                    new_messages=[{"role": "user", "content": user_message}],
+                )
```
[the remark this answers is not in the document]
Comment (holdenk) [reply fighthealthinsurance/chat_interface.py:1230]: Good catch — fixed in 8c7d36e: the end-of-turn persists (success and failure paths) now resubmit the user message only when the pre-persist failed, so a concurrent turn's merge into the pending message can no longer cause the "A B" → "A B A" re-append. Added a regression test that races a second pre-persist mid-generation.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: fighthealthinsurance/ml/ml_models.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (holdenk) [reply fighthealthinsurance/ml/ml_models.py:1387]: Adopted in 7261dcd: `_call_llm_with_actions` derives `allow_repeated_reply` from the raw user message (before prompt wrapping) and threads it through `build_llm_calls` / `build_llm_calls_for_variants` / `retry_llm_with_fallback` into `generate_chat_response`, where `_repeats_last_reply` now returns False when the user asked for the repeat — so an explicit repeat request no longer burns the serial corrective retry, while loop detection stays intact for wrapped/injected prompt text. Covered by tests for the flag reaching backends via the retry path and for the self-heal loop delivering a requested repeat in a single inference call.

---
_Generated by [Claude Code](https://claude.ai/code)_
